x-pack/plugin/core/src/main/java/org/elasticsearch/xpack/core/ilm/GenerateRollupIndexNameStep.java

```diff
@@ -0,0 +1,80 @@
+ * 2.0.
+ */
+package org.elasticsearch.xpack.core.ilm;
+
+import org.apache.logging.log4j.LogManager;
+import org.apache.logging.log4j.Logger;
+import org.elasticsearch.cluster.ClusterState;
+import org.elasticsearch.cluster.metadata.IndexMetadata;
+import org.elasticsearch.cluster.metadata.Metadata;
+import org.elasticsearch.common.UUIDs;
+import org.elasticsearch.index.Index;
+
+import java.util.Locale;
+import java.util.Objects;
+
+import static org.elasticsearch.xpack.core.ilm.LifecycleExecutionState.ILM_CUSTOM_METADATA_KEY;
+import static org.elasticsearch.xpack.core.ilm.LifecycleExecutionState.fromIndexMetadata;
+
+/**
+ * Generates a new rollup index name for the current managed index.
+ * <p>
+ * The generated rollup index name will be in the format {rollup-indexName-randomUUID}
+ * eg.: rollup-myindex-cmuce-qfvmn_dstqw-ivmjc1etsa
+ */
+public class GenerateRollupIndexNameStep extends ClusterStateActionStep {
+
+    public static final String NAME = "generate-rollup-name";
+
+    private static final Logger logger = LogManager.getLogger(GenerateRollupIndexNameStep.class);
+
+    public GenerateRollupIndexNameStep(StepKey key, StepKey nextStepKey) {
+        super(key, nextStepKey);
+    }
+
+    @Override
+    public ClusterState performAction(Index index, ClusterState clusterState) {
+        IndexMetadata indexMetaData = clusterState.metadata().index(index);
+        if (indexMetaData == null) {
+            // Index must have been since deleted, ignore it
+            logger.debug("[{}] lifecycle action for index [{}] executed but index no longer exists", getKey().getAction(), index.getName());
+            return clusterState;
+        }
+        ClusterState.Builder newClusterStateBuilder = ClusterState.builder(clusterState);
+        LifecycleExecutionState lifecycleState = fromIndexMetadata(indexMetaData);
+        assert lifecycleState.getRollupIndexName() == null : "index " + index.getName() +
+            " should have a rollup index name generated by the ilm policy but has " + lifecycleState.getRollupIndexName();
+        LifecycleExecutionState.Builder newCustomData = LifecycleExecutionState.builder(lifecycleState);
+        String rollupIndexName = generateRollupIndexName(index.getName());
+        newCustomData.setRollupIndexName(rollupIndexName);
+        IndexMetadata.Builder indexMetadataBuilder = IndexMetadata.builder(indexMetaData);
+        indexMetadataBuilder.putCustom(ILM_CUSTOM_METADATA_KEY, newCustomData.build().asMap());
+        newClusterStateBuilder.metadata(Metadata.builder(clusterState.getMetadata()).put(indexMetadataBuilder));
+        return newClusterStateBuilder.build();
+    }
+
+    // TODO(talevy): find alternative to lowercasing UUID? kind of defeats the expectation of the UUID here. index names must lowercase
+    public static String generateRollupIndexName(String indexName) {
+        return "rollup-" + indexName + "-" +  UUIDs.randomBase64UUID().toLowerCase(Locale.ROOT);
```

Comment:
I agree, this is a strange departure. I believe Shrink Action in ILM did not use the `.` syntax because ILM operates more generally on indices, not data streams. So `shrunk-` as a prefix made sense. For consistency (https://github.com/elastic/elasticsearch/pull/65633#discussion_r552879406) with both Searchable Snapshot and Shrink, this `rollup-` prefix was chosen.

If we wish to change this depending on whether ILM is operating on a datastream or not, I think we can bring this up for discussion with the ILM team.

Reply:
I'd like to continue this conversation outside of this PR if you think that is reasonable.

It may make sense to update all the actions to follow this pattern.

Reply:
I think this is part of a bigger discussion (outside of this PR), we can handle it on the core/features team and we can update this if/when we have a decision on how to treat these

Reply:
In this case, I am fine with moving on with this naming now. Thanks

Reply:
thank you everyone!